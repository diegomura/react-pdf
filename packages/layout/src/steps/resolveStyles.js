import * as P from '@react-pdf/primitives';
import stylesheet from '@react-pdf/stylesheet';

const isLink = node => node.type === P.Link;

const DEFAULT_LINK_STYLES = {
  color: 'blue',
  textDecoration: 'underline',
};

/**
 * Computes styles using stylesheet
 *
 * @param {Object} container
 * @param {Object} document node
 * @returns {Object} computed styles
 */
const computeStyle = (container, node) => {
  let baseStyle = node.style;

  if (isLink(node)) {
    baseStyle = Array.isArray(node.style)
      ? [DEFAULT_LINK_STYLES, ...node.style]
      : [DEFAULT_LINK_STYLES, node.style];
  }

  return stylesheet(container, baseStyle);
};

/**
 * Resolves node styles
 *
 * @param {Object} container
 * @param {Object} document node
 * @returns {Object} node (and subnodes) with resolved styles
 */
const resolveNodeStyles = container => node => {
  const style = computeStyle(container, node);

  if (!node.children) return Object.assign({}, node, { style });

  const children = node.children.map(resolveNodeStyles(container));

  return Object.assign({}, node, { style, children });
};

/**
 * Resolves page styles
 *
 * @param {Object} document page
 * @returns {Object} document page with resolved styles
 */
const resolvePageStyles = page => {
  const dpi = page.props?.dpi || 72;
  const width = page.box?.width || page.style.width;
  const height = page.box?.height || page.style.height;
  const orientation = page.props?.orientation || 'portrait';
  const container = { width, height, orientation, dpi };

  return resolveNodeStyles(container)(page);
};

/**
 * Resolves document styles
 *
 * @param {Object} document root
 * @returns {Object} document root with resolved styles
 */
const resolveStyles = root => {
  if (!root.children) return root;

  const children = root.children.map(resolvePageStyles);

  return Object.assign({}, root, { children });
};

export default resolveStyles;
