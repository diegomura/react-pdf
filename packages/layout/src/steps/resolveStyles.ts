import * as P from '@react-pdf/primitives';
import stylesheet, { Container, Style } from '@react-pdf/stylesheet';

import {
  DocumentNode,
  Node,
  PageNode,
  SafeDocumentNode,
  SafeNode,
} from '../types';

const isLink = (node: Node) => node.type === P.Link;

const DEFAULT_LINK_STYLES: Style = {
  color: 'blue',
  textDecoration: 'underline',
};

/**
 * Computes styles using stylesheet
 *
 * @param container
 * @param node - Document node
 * @returns Computed styles
 */
const computeStyle = (container: Container, node: Node) => {
  let baseStyle: Style[] = [node.style as Style];

  if (isLink(node)) {
    baseStyle = Array.isArray(node.style)
      ? [DEFAULT_LINK_STYLES, ...node.style]
      : [DEFAULT_LINK_STYLES, node.style];
  }

  return stylesheet(container, baseStyle);
};

/**
 * Resolves node styles
 *
 * @param container
 * @returns Resolve node styles
 */
const resolveNodeStyles =
  (container: Container) =>
  (node: Node): SafeNode => {
    const style = computeStyle(container, node);

    if (!node.children) return Object.assign({}, node, { style }) as SafeNode;

    const children = node.children.map(resolveNodeStyles(container));

    return Object.assign({}, node, { style, children }) as SafeNode;
  };

/**
 * Resolves page styles
 *
 * @param page Document page
 * @returns Document page with resolved styles
 */
export const resolvePageStyles = (page: PageNode) => {
  const dpi = page.props?.dpi || 72;
  const style = page.style as Style;
  const width = page.box?.width || style.width;
  const height = page.box?.height || style.height;
  const orientation = page.props?.orientation || 'portrait';
  const remBase = style?.fontSize || 18;
  const container = { width, height, orientation, dpi, remBase } as Container;

  return resolveNodeStyles(container)(page);
};

/**
 * Resolves document styles
 *
 * @param root - Document root
 * @returns Document root with resolved styles
 */
const resolveStyles = (root: DocumentNode): SafeDocumentNode => {
  if (!root.children) return root as SafeDocumentNode;

  const children = root.children.map(resolvePageStyles);

  return Object.assign({}, root, { children }) as SafeDocumentNode;
};

export default resolveStyles;
