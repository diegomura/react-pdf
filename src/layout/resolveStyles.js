import * as R from 'ramda';

import isLink from '../node/isLink';
import flattenStyles from '../stylesheet/flatten';
import expandStyles from '../stylesheet/expandStyles';
import transformUnits from '../stylesheet/transformUnits';
import transformStyles from '../stylesheet/transformStyles';
import transformColors from '../stylesheet/transformColors';
import resolveMediaQueries from '../stylesheet/resolveMediaQueries';

const LINK_STYLES = {
  color: 'blue',
};

/**
 * Filter styles with `none` value
 *
 * @param {Object} style object
 * @returns {Object} style without none values
 */
const filterNoneValues = R.reject(R.equals('none'));

/**
 * Resolves styles
 *
 * @param {Object} container
 * @param {Object} node
 * @param {Object} style object
 * @returns {Object} resolved style object
 */
const resolveStyles = container =>
  R.compose(
    transformUnits(container),
    transformColors,
    transformStyles,
    expandStyles,
    resolveMediaQueries(container),
    filterNoneValues,
    flattenStyles,
  );

/**
 * Resolves node styles
 *
 * @param {Object} container
 * @param {Object} document node
 * @returns {Object} node (and subnodes) with resolved styles
 */
const resolveNodeStyles = container => node =>
  R.o(
    R.when(isLink, R.evolve({ style: R.merge(LINK_STYLES) })),
    R.evolve({
      style: resolveStyles(container),
      children: R.map(resolveNodeStyles(container)),
    }),
  )(node);

/**
 * Resolves page styles
 *
 * @param {Object} document page
 * @returns {Object} document page with resolved styles
 */
const resolvePageStyles = page => {
  const box = R.prop('box', page);
  const style = R.prop('style', page);
  const container = R.isEmpty(box) ? style : box;

  return R.evolve({
    style: resolveStyles(container),
    children: R.map(resolveNodeStyles(container)),
  })(page);
};

/**
 * Resolves root styles
 *
 * @param {Object} document root
 * @returns {Object} document root with resolved styles
 */
export default R.evolve({
  children: R.map(resolvePageStyles),
});
