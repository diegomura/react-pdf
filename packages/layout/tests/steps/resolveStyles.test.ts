import { describe, expect, test } from 'vitest';

import resolveStyles from '../../src/steps/resolveStyles';

describe('layout resolveStyles', () => {
  test('Should resolve page styles', () => {
    const result = resolveStyles({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          style: {
            paddingHorizontal: '10in',
            paddingVertical: '10mm',
            backgroundColor: 'red',
            border: '1cm dotted green',
          },
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should resolve several pages styles', () => {
    const result = resolveStyles({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          style: {
            paddingHorizontal: '10in',
            paddingVertical: '10mm',
            borderTop: '4 solid #FF00000',
          },
        },
        {
          type: 'PAGE',
          props: {},
          style: {
            backgroundColor: 'red',
            border: '1cm dotted green',
            fontWeight: 'bold',
          },
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should resolve page styles array', () => {
    const result = resolveStyles({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          style: [
            {
              paddingHorizontal: '10in',
              paddingVertical: '10mm',
            },
            {
              backgroundColor: 'red',
              border: '1cm dotted green',
            },
          ],
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should resolve nested node styles', () => {
    const result = resolveStyles({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          box: {
            top: 0,
            right: 0,
            bottom: 0,
            left: 0,
            width: 100,
            height: 200,
          },
          children: [
            {
              type: 'VIEW',
              props: {},
              style: {
                paddingHorizontal: '10in',
                paddingVertical: '10mm',
                backgroundColor: 'red',
                border: '1cm dotted green',
              },
            },
          ],
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should resolve nested node styles media queries', () => {
    const result = resolveStyles({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          box: {
            top: 0,
            right: 0,
            bottom: 0,
            left: 0,
            width: 100,
            height: 200,
          },
          children: [
            {
              type: 'VIEW',
              props: {},
              style: {
                backgroundColor: 'red',
                '@media max-width: 500': {
                  backgroundColor: 'green',
                },
              },
            },
            {
              type: 'VIEW',
              props: {},
              style: {
                backgroundColor: 'red',
                '@media min-width: 500': {
                  backgroundColor: 'green',
                },
              },
            },
          ],
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should resolve nested node styles media queries with page style', () => {
    const result = resolveStyles({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          style: { width: 100, height: 200 },
          children: [
            {
              type: 'VIEW',
              props: {},
              style: {
                backgroundColor: 'red',
                '@media max-width: 500': {
                  backgroundColor: 'green',
                },
              },
            },
            {
              type: 'VIEW',
              props: {},
              style: {
                backgroundColor: 'red',
                '@media min-width: 500': {
                  backgroundColor: 'green',
                },
              },
            },
          ],
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should resolve nested node styles array', () => {
    const result = resolveStyles({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          box: {
            top: 0,
            right: 0,
            bottom: 0,
            left: 0,
            width: 100,
            height: 200,
          },
          children: [
            {
              type: 'VIEW',
              props: {},
              style: [
                {
                  paddingHorizontal: '10in',
                  paddingVertical: '10mm',
                },
                {
                  backgroundColor: 'red',
                  border: '1cm dotted green',
                },
              ],
            },
          ],
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should resolve default link styles', () => {
    const result = resolveStyles({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          box: {
            top: 0,
            right: 0,
            bottom: 0,
            left: 0,
            width: 100,
            height: 200,
          },
          children: [
            {
              type: 'LINK',
              props: {},
              style: {},
            },
          ],
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should overide default link styles', () => {
    const result = resolveStyles({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          box: {
            top: 0,
            right: 0,
            bottom: 0,
            left: 0,
            width: 100,
            height: 200,
          },
          children: [
            {
              type: 'LINK',
              props: {},
              style: { color: 'wheat', textDecoration: 'none' },
            },
          ],
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });

  test('Should overide default link styles with array', () => {
    const result = resolveStyles({
      type: 'DOCUMENT',
      props: {},
      children: [
        {
          type: 'PAGE',
          props: {},
          box: {
            top: 0,
            right: 0,
            bottom: 0,
            left: 0,
            width: 100,
            height: 200,
          },
          children: [
            {
              type: 'LINK',
              props: {},
              style: [{ color: 'wheat', textDecoration: 'none' }],
            },
          ],
        },
      ],
    });

    expect(result).toMatchSnapshot();
  });
});
