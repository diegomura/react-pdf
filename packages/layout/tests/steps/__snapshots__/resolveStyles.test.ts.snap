// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`layout resolveStyles > Should overide default link styles 1`] = `
{
  "children": [
    {
      "box": {
        "height": 200,
        "width": 100,
      },
      "children": [
        {
          "props": {},
          "style": {
            "color": "wheat",
            "textDecoration": "none",
          },
          "type": "LINK",
        },
      ],
      "props": {},
      "style": {},
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolveStyles > Should overide default link styles with array 1`] = `
{
  "children": [
    {
      "box": {
        "height": 200,
        "width": 100,
      },
      "children": [
        {
          "props": {},
          "style": {
            "color": "wheat",
            "textDecoration": "none",
          },
          "type": "LINK",
        },
      ],
      "props": {},
      "style": {},
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolveStyles > Should resolve default link styles 1`] = `
{
  "children": [
    {
      "box": {
        "height": 200,
        "width": 100,
      },
      "children": [
        {
          "props": {},
          "style": {
            "color": "blue",
            "textDecoration": "underline",
          },
          "type": "LINK",
        },
      ],
      "props": {},
      "style": {},
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolveStyles > Should resolve nested node styles 1`] = `
{
  "children": [
    {
      "box": {
        "height": 200,
        "width": 100,
      },
      "children": [
        {
          "props": {},
          "style": {
            "backgroundColor": "red",
            "borderBottomColor": "green",
            "borderBottomStyle": "dotted",
            "borderBottomWidth": 28.346456692913385,
            "borderLeftColor": "green",
            "borderLeftStyle": "dotted",
            "borderLeftWidth": 28.346456692913385,
            "borderRightColor": "green",
            "borderRightStyle": "dotted",
            "borderRightWidth": 28.346456692913385,
            "borderTopColor": "green",
            "borderTopStyle": "dotted",
            "borderTopWidth": 28.346456692913385,
            "paddingBottom": 28.346456692913385,
            "paddingLeft": 720,
            "paddingRight": 720,
            "paddingTop": 28.346456692913385,
          },
          "type": "VIEW",
        },
      ],
      "props": {},
      "style": {},
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolveStyles > Should resolve nested node styles array 1`] = `
{
  "children": [
    {
      "box": {
        "height": 200,
        "width": 100,
      },
      "children": [
        {
          "props": {},
          "style": {
            "backgroundColor": "red",
            "borderBottomColor": "green",
            "borderBottomStyle": "dotted",
            "borderBottomWidth": 28.346456692913385,
            "borderLeftColor": "green",
            "borderLeftStyle": "dotted",
            "borderLeftWidth": 28.346456692913385,
            "borderRightColor": "green",
            "borderRightStyle": "dotted",
            "borderRightWidth": 28.346456692913385,
            "borderTopColor": "green",
            "borderTopStyle": "dotted",
            "borderTopWidth": 28.346456692913385,
            "paddingBottom": 28.346456692913385,
            "paddingLeft": 720,
            "paddingRight": 720,
            "paddingTop": 28.346456692913385,
          },
          "type": "VIEW",
        },
      ],
      "props": {},
      "style": {},
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolveStyles > Should resolve nested node styles media queries 1`] = `
{
  "children": [
    {
      "box": {
        "height": 200,
        "width": 100,
      },
      "children": [
        {
          "props": {},
          "style": {
            "backgroundColor": "green",
          },
          "type": "VIEW",
        },
        {
          "props": {},
          "style": {
            "backgroundColor": "red",
          },
          "type": "VIEW",
        },
      ],
      "props": {},
      "style": {},
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolveStyles > Should resolve nested node styles media queries with page style 1`] = `
{
  "children": [
    {
      "children": [
        {
          "props": {},
          "style": {
            "backgroundColor": "green",
          },
          "type": "VIEW",
        },
        {
          "props": {},
          "style": {
            "backgroundColor": "red",
          },
          "type": "VIEW",
        },
      ],
      "props": {},
      "style": {
        "height": 200,
        "width": 100,
      },
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolveStyles > Should resolve page styles 1`] = `
{
  "children": [
    {
      "props": {},
      "style": {
        "backgroundColor": "red",
        "borderBottomColor": "green",
        "borderBottomStyle": "dotted",
        "borderBottomWidth": 28.346456692913385,
        "borderLeftColor": "green",
        "borderLeftStyle": "dotted",
        "borderLeftWidth": 28.346456692913385,
        "borderRightColor": "green",
        "borderRightStyle": "dotted",
        "borderRightWidth": 28.346456692913385,
        "borderTopColor": "green",
        "borderTopStyle": "dotted",
        "borderTopWidth": 28.346456692913385,
        "paddingBottom": 28.346456692913385,
        "paddingLeft": 720,
        "paddingRight": 720,
        "paddingTop": 28.346456692913385,
      },
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolveStyles > Should resolve page styles array 1`] = `
{
  "children": [
    {
      "props": {},
      "style": {
        "backgroundColor": "red",
        "borderBottomColor": "green",
        "borderBottomStyle": "dotted",
        "borderBottomWidth": 28.346456692913385,
        "borderLeftColor": "green",
        "borderLeftStyle": "dotted",
        "borderLeftWidth": 28.346456692913385,
        "borderRightColor": "green",
        "borderRightStyle": "dotted",
        "borderRightWidth": 28.346456692913385,
        "borderTopColor": "green",
        "borderTopStyle": "dotted",
        "borderTopWidth": 28.346456692913385,
        "paddingBottom": 28.346456692913385,
        "paddingLeft": 720,
        "paddingRight": 720,
        "paddingTop": 28.346456692913385,
      },
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;

exports[`layout resolveStyles > Should resolve several pages styles 1`] = `
{
  "children": [
    {
      "props": {},
      "style": {
        "borderTopColor": "#FF00000",
        "borderTopStyle": "solid",
        "borderTopWidth": 4,
        "paddingBottom": 28.346456692913385,
        "paddingLeft": 720,
        "paddingRight": 720,
        "paddingTop": 28.346456692913385,
      },
      "type": "PAGE",
    },
    {
      "props": {},
      "style": {
        "backgroundColor": "red",
        "borderBottomColor": "green",
        "borderBottomStyle": "dotted",
        "borderBottomWidth": 28.346456692913385,
        "borderLeftColor": "green",
        "borderLeftStyle": "dotted",
        "borderLeftWidth": 28.346456692913385,
        "borderRightColor": "green",
        "borderRightStyle": "dotted",
        "borderRightWidth": 28.346456692913385,
        "borderTopColor": "green",
        "borderTopStyle": "dotted",
        "borderTopWidth": 28.346456692913385,
        "fontWeight": 700,
      },
      "type": "PAGE",
    },
  ],
  "props": {},
  "type": "DOCUMENT",
}
`;
